src/core/qgstextrenderer.cpp

```diff
@@ -1925,25 +1926,24 @@ void QgsTextFormat::readXml( const QDomElement &elem, const QgsReadWriteContext
   {
     mShadowSettings.readXml( textStyleElem );
   }
-  if ( textStyleElem.firstChildElement( QStringLiteral( "background" ) ).isNull() )
-  {
-    mBackgroundSettings.readXml( elem, context );
-  }
-  else
-  {
-    mBackgroundSettings.readXml( textStyleElem, context );
-  }
 
-  if ( textStyleElem.firstChildElement( QStringLiteral( "dd_properties" ) ).isNull() )
+  QDomElement ddElem = textStyleElem.firstChildElement( QStringLiteral( "dd_properties" ) );
+  if ( ddElem.isNull() )
```

Comment:
Hmm - this doesn't look right. We'd still want to restore these even when data defined properties are present. What's the logic here?

Reply:
> properties are present. What's the logic here?

I said it wasn't clear to me :)

The question is what is the current (broken) logic ? 

In case of missing DD and issue #31427, the QML is read twice:
First time here:
https://github.com/qgis/QGIS/blob/master/src/core/qgstextrenderer.cpp#L1934
which gets the correct config
and second time here:
https://github.com/qgis/QGIS/blob/master/src/core/qgstextrenderer.cpp#L1939
which gets the wrong config

The DD getter also seems to be duplicated:
here is checked intially https://github.com/qgis/QGIS/blob/master/src/core/qgstextrenderer.cpp#L1937
and here is checked again: https://github.com/qgis/QGIS/blob/master/src/core/qgstextrenderer.cpp#L1946
I moved the getter before to simplify the code.